### ISSUE TITLE
Gradle Daemon created for TAPI request gets unexpected working directory

### ISSUE THREAD

**Issue by vladsoroka** [gradle/gradle#16748]
Gradle Daemon started for Gradle TAPI operation uses working directory of the TAPI client process although `org.gradle.tooling.GradleConnector#forProjectDirectory` has been set.  

Yes, it's a bad practice to have CWD dependant code in a build scripts. But the current behaviour is inconsistent with the terminal user experience and this leads to confusion.

### Expected Behavior
The same behaviour for TAPI and terminal Gradle running.
For terminal run the newly created Gradle daemon is started with current terminal working directry.
For TAPI run the newly created Gradle daemon should be started using `GradleConnector#forProjectDirectory` value.

### Current Behavior
For TAPI run the newly created Gradle daemon is started using IDE cwd.

### Context
This report is not about changing of CWD for existing Gradle daemons which can be reused if it's compatible.
It's only about newly created Daemon processes, so the behaviour will become consistent with the command line usages.

The origin user report and other users complains can be found at https://youtrack.jetbrains.com/issue/IDEA-265203 and linked issues.

**Comment by Vampire** [comment]
This also causes `generatePrecompiledScriptPluginAccessors` to fail as it tries to create `.gradle\vcs-1` within the IntelliJ installation directory. So even if newly created daemons would get a project directory as CWD, it would probably still not be the correct one when running mutliple builds on that daemon.

**Comment by oehme** [comment]
@Vampire I'm pretty sure that's unrelated. Gradle uses the project directory for this. Sounds more like IDEA is not passing the correct projectDir to the Tooling API.

@vladsoroka CWD is undefined in the Gradle daemon (we set it to some folder inside the Gradle user home). I'm not sure why it would work on the command line, but it doesn't really matter, it is not to be relied on. Build code must never depend on it. The problem is in the plugins that the reporters of that issue are using. That plugin should use `projectDir` instead of relying on CWD.

**Comment by oehme** [comment]
We should stop even trying to set the CWD in the daemon, as it can't be reliably done anyway. Build logic should use `Project.file` and friends.

@Vampire that sounds more like something that needs to be fixed in composite builds. Seems like they depend on CWD, which they shouldn't.

**Comment by oehme** [comment]
After some more investigation it turns out that there was indeed a bug in how `StartParameter#currentDir` was set up in the TAPI. I suppose this never came up in the integration tests, because they were just happily writing their caches to an unexpected, but writable location. I've fixed that in #16755 and adjusted the unit tests.

This also "fixes" the build scripts/plugins relying on CWD, but again, I think we should stop doing this magic :)

**Comment by big-guy** [comment]
@oehme does it look like this changed recently?  

Is there a way to fix this on older Gradle versions with a change on the consumer TAPI side?

**Comment by vladsoroka** [comment]
@oehme 
>CWD in the daemon is set to the CWD of the client. On the CLI that's whatever folder your terminal is currently in. In the IDE that's the CWD of IDEA.

Yes, that's what I've described as the current behavior at the issue description.
> This is expected.

I can not agree :). It's not expected because CLI user has different behaviour. Also a number of users complains confirm this.
>we should stop even trying to set the CWD in the daemon, as it is meaningless.

Yeah! But it should be done both for tooling API and CLI clients

Thanks for the fix!

**Comment by oehme** [comment]
@big-guy no, this was always broken. The JetBrains team worked around this by changing the user.dir of IntelliJ. They could reintroduce that change, but as you might imagine, this causes problems for other parts of the IDE. So ideally we'd do a hotfix for this, so 2021.1 can be fixed.

**Comment by vladsoroka** [comment]
>Is there a way to fix this on older Gradle versions with a change on the consumer TAPI side?

@oehme @big-guy +100, we need this fix on TAPI side. At least for target Gradle versions which supports precompiled script plugins

**Comment by oehme** [comment]
I believe this also causes https://github.com/gradle/gradle/issues/6825 and https://github.com/gradle/gradle/issues/8353.

**Comment by nskvortsov** [comment]
The current dir problem hits users not only in `build.gradle` files, but also in `settings.gradle`: https://youtrack.jetbrains.com/issue/IDEA-266253

Feels like almost any invocation of `new File()` should produce a warning

**Comment by oehme** [comment]
> Feels like almost any invocation of new File() should produce a warning

That would be very hard to get right, but we could set the CWD of the daemon to a directory called `dont-rely-on-cwd` ;)

**Comment by nskvortsov** [comment]
Actually, I'd say that making daemons CWD stable and pointing to something like `GRADE_USER_HOME/daemon_id/workdir` would have been a good decision - users would not rely on CWD from the very beginning.

**Comment by oehme** [comment]
Yes, we do that for the worker API already and I'm advocating for doing the same for the main daemon. Mutating the current dir also doesn't fully work. e.g. `java.nio.Path` will keep reporting the initial dir.

**Comment by nskvortsov** [comment]
Now it is too late, as there are a lot of scripts relying on (occasional) CWD pointing to project directory.

**Comment by oehme** [comment]
It's not too late, we can do that change in the next major release. All these usages are broken and probably causing weird bugs for the people who have them in their scripts.

**Comment by oehme** [comment]
Another problem this surfaced is that the Tooling API wrapper download uses the wrong working directory. I'll adjust that as well.

@big-guy Heads-up, we'll definitely need a 6.8.x release to unblock JetBrains and fix IntelliJ 2021.1

**Comment by oehme** [comment]
Unfortunately, my initial investigation was a dead-end. Turns out that the StartParameter code path is correct after all, even though the naming is misleading. The CWD issue is instead created by [this call ](https://github.com/gradle/gradle/blob/master/subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/DaemonBuildActionExecuter.java#L45)to `SystemProperties#getCurrentDir`. This code is on the "provider" side of the Tooling API, i.e. inside the target Gradle version. So even if we changed this, it would not change IDEA for older Gradle versions.

Right now I'm out of ideas on how to fix the general CWD issue for older Gradle versions. The best I can think of is fixing the sub-issues that are not the user's fault, i.e. #12060 (which I have a PR for) and https://youtrack.jetbrains.com/issue/IDEA-266067 (which I can't reproduce yet).

**Comment by oehme** [comment]
@vladsoroka @nskvortsov Would be possible for you to add `System.setProperty('user.dir', gradle.startParameter.currentDir)` to your init script that you inject into every build? This would solve all the issues except for the wrapper download issue, which will be fixed by https://github.com/gradle/gradle/pull/16755

**Comment by vladsoroka** [comment]
@oehme [we've added the old workaround](https://github.com/JetBrains/intellij-community/commit/5536877085dbefc10e0526f382626b393fa83a22) (with additional condition to avoid known breakage of another IntelliJ subsystem) for IDEA 2021.1.1.
Thank you for the support, I hope we will be able to sort it out (w/o changing IDE `user.dir`) till the next release.

**Comment by oehme** [comment]
@vladsoroka so what do you think about my proposal above?

**Comment by vladsoroka** [comment]
@oehme changing `user.dir` of the running Gradle daemon process is also looks like a workaround. It will not affect IDEA process but might lead to further issues/confusions.

Ideally, I would try to find a way to use `projectDir` as CWD when a new Daemon java process is being created by the TAPI request. Maybe we can try wrap code responsible for the process creation on TAPI side in a target build Gradle version independent way?

**Comment by oehme** [comment]
@vladsoroka Setting user.dir to the project dir is exactly what Gradle does when invoked from the command line, so doing the same in an init scrpt it would not lead to other problems. It is the appropriate fix given the situation. Trying to fix up the TAPI would be way more complicated and would not give you any better result.

**Comment by vladsoroka** [comment]
>It is then changed at the start of every build to match the CWD of the client.
I didn't know this. Thanks for the link. It means the results of `System.getProperty('user.dir')` and `new File("").absolutePath` will be different (on Java 11+) both for TAPI and CLI. It's also can be confusing but possibly acceptable.

@big-guy what do you think? should it be fixed this way for TAPI clients? Will it work fine with regards to other Gradle features like configuration cache? How does it work in Buildship?

**Comment by oehme** [comment]
Buildship never had the workaround that IDEA had, so projects that use relative paths simply don't work in Buildship. They also don't work correctly in Java 11, as you already pointed out. That's why I said that Gradle 8 should remove this behavior entirely.

**Comment by jjohannes** [comment]
I haven't looked at what has been discussed here before, so please bear with me if I misunderstand anything.

I think the title of this issue is misleading, as the working dir the daemon gets when it is spawned is not the issue it seems. Instead, changing the dir when a build is started was inconsistent but is now fixed for 7.1 with #16755. If my understanding is correct, IDEA could remove the much discussed "workaround" if it updates the Tooling API client to 7.1 then.

Is my understanding correct?

If yes, I think this issue can be closed.

If not, please share a pointer to what else shall be done on the Gradle side (if this was already said, please give me a pointer).

**Comment by oehme** [comment]
@jjohannes no, you misunderstood. As I said in #16755, that pull request does not do anything about the daemon's working dir. It only fixes the working dir and system properties of the wrapper download, nothing else.

The daemon working dir can't be fixed, as that logic lives in already released Gradle versions. I think we should completely remove this behavior, as it just hides bugs in people's build scripts. The working dir should always be somewhere constant like the Gradle user home. We already do this for the worker API.

**Comment by eskatos** [comment]
Another reason to set CWD outside of the build root is that on Windows the daemon working directory is locked. See
* https://github.com/gradle/gradle/issues/26447

**Comment by VictorPascualV** [comment]
⚠️ We're also blocked by this issue.

We're using Gradle Tooling API embedded in a backend service, that when deployed, it only has write permissions to a particular folder. We set `useGradleUserHomeDir()` with a directory within the folder with write permissions. But Gradle Tooling API still tries to start the Daemon in `'user.dir'`, and fails, because that folder has read-only permissions.

Are there any plans to fix this issue? If not, is there any workaround that we could use?

Thanks!

**Comment by agentgt** [comment]
Hey guys I noticed unusually behavior with Gradle and CWD. I'm not an expert on Gradle but it appears Gradle actually resets the CWD unlike Maven or Maven Daemon.

Does Gradle uses native code to do this?

I will copy my comments on a bug in my project:

----

It appears that Gradle fundamentally does bizarre things with the CWD.

The CWD in gradle is always the daemon. However it is not the true CWD where the gradle command was issued.

However the parts of the java.nio system maintain the correct CWD.

If you do in the annotation processor Path.of(".").toAbsolutePath() it will show some gradle daemon path. However if you do Files.newInputStream(Path.of("./somefile")) it will resolve and be able to find the file assuming somefile is in the CWD where the gradle command is issued on MacOS and Linux but not Windows.

This really bizarre behavior that makes me thing CWD is just never the right choice. It was always the fallback but given how unreliable it is perhaps we should not even fallback to it.

Probably the better solution is to use annotation processor Filer path location and go up the directories looking for a build.gradle or build.gradle.kts like we do with Eclipse for .classpath.

Furthermore it is strange how Maven and the Maven Daemon do not have this problem and actually resolve to the correct CWD the command was issued in. So Gradle in its script must be resetting the CWD (e.g. cd) OR it has some native code doing it and the java.nio is ignoring that resetting of the CWD.

**Comment by Vampire** [comment]
Yes, relying on CWD in Gradle logic is almost always the wrong thing to do, as in practically any JVM program.
The only case I'm aware, where it is appropriate to use the CWD in JVM programs is, if you use a relative path given by the callling user on the commandline.
If you for example have some application (independent of Gradle) that uses `new File(hardcodedPathRelativeToStartBatFile)`, this also only works if the user calls that bat file from the expected directory, but not if he for example calls it from another directory.

In the case of code running in the context of a Gradle build, this is even "worse", as the code is usually run within the Gradle daemon which can be reused among different projects that can have different CWDs when you invoke the CLI.
Gradle tries to set the CWD to the root project directory (or maybe the CLI CWD, I don't remember) in an already running daemon, but as you have seen, this does not always fully work or does not work for all API as some determine the CWD once and remember it as there is no really supported way to change the CWD in-process.
Gradle just tries its best to set the CWD as good as it can to support "broken" code that relies on the CWD.

Gradle logic should usually always resolve paths relative to the project directory or similar.

**Comment by agentgt** [comment]
@Vampire yes I would never use the CWD normally. I'm in the context of an annotation processor where there are possibly 3 implementations of what is called the `Filer` (Eclipse, javac, and Gradle kind of).

It would take me a long time go over how the only thing supported in terms of paths is `javax.tools.StandardLocation#CLASS_OUTPUT` and how folks do not always put resources and or have a build system that will always copy to it.

So if an annotation processor tries to access things that are not java code (in my case mustache templates) it is a pain the butt. CWD was used as a last resort but now I will probably just fail the build and require absolute path passed to the annotation processor.

https://jstach.io/doc/jstachio/current/apidocs/io.jstach.jstache/io/jstach/jstache/JStache.html#RESOURCES_PATH_OPTION

(ignore the term CWD in that doc).

Ideally some system property or something would be set that is standard across the build systems of what the current project directory is but that probably is not going to happen and or there probably issues with that as well (especially if forking is not happening).

**Comment by Vampire** [comment]
Long ago that I wrote my last annotation processor, but yeah, giving that in as property and requiring that the right path is given sounds like the most reliable way to me.

**Comment by eskatos** [comment]
We've sketched a plan to finally resolve this and would like feedback before committing, especially from TAPI consumers and anyone whose build relies on the daemon's working directory.

The direction matches the consensus in this thread: stop changing the daemon's working directory per build. The daemon will use a **stable working directory** inside the Gradle user home, the same approach the Worker API already uses, instead of inheriting the client's directory (your terminal dir on the CLI, the IDE process dir via the Tooling API).

In a Gradle minor version (no breaking change by default)
- Builds keep working as today, but emit a deprecation warning where we detect build logic resolving a relative path against the working directory (i.e. code that will break).
- A new opt-in property, `org.gradle.daemon.stable-working-dir=true`, switches the daemon to the full future behaviour so you can test your build/plugins ahead of time.

In the next Gradle major version (the change)
- The stable working directory becomes the default; relative-path reliance in build logic fails loudly with an actionable message pointing at `ProjectLayout` / `BuildLayout`.
- The opt-in property becomes a no-op and is deprecated.
- This lands alongside the configuration cache becoming the default, which already discourages CWD reliance. The two reinforce each other.

The warning is feasible now. Earlier in this thread, accurately detecting CWD reliance was (rightly) considered impractical: a blanket `new File()` warning would be far too noisy. That has changed. The configuration cache work has since built a robust call-interception / instrumentation layer for build logic (scripts, plugins, buildSrc), and it's active even when the configuration cache itself is off. We can reuse it to spot *only* the relevant case, a relative path actually resolved against the working directory, rather than warning on every file access. This infrastructure simply didn't exist when this issue was opened.

There would be some limitations: the warning/fail-loud covers instrumented build logic (scripts, plugins, buildSrc). Code we can't instrument (the JDK, third-party libraries loaded outside the build classpath, forked processes) won't always be caught; relative reads there will fail naturally, but stray relative writes could still be silent.

Before finalising, we would like to run the opt-in flag across popular plugins, IDE/Tooling-API scenarios, to measure impact and tune the migration. We would ship the opt-in flag alone at first so that if your build or plugin depends on the daemon's working directory today, you could try the flag once it's available, and report what breaks.

**Comment by jjohannes** [comment]
The proposal sounds very good!
I share one use case where we were tempted to use `new File()` but found another solution that may not be obvious:

-----

In a **composite build** setup you sometimes need to find the root of the **build tree**. You can do this as described in #14697.

```kotlin
fun Gradle.rootBuild(): Gradle = parent.let { it?.rootBuild() ?: this }

gradle.rootBuild() // do something only to the root
```

If you need to access a file in the root (e.g. `local.properties`), you can do it via `rootProject`, for example:

```
gradle.rootBuild().rootProject.rootDir
```

However, if you need to do this early from the settings file `rootProject` mayan be inaccessible as Gradle has not yet evaluated the projects. Then there is, as far as I know, no way to access `rootDir`. The recently introduced `settings.layout` cannot be used as there is no way to get there from `gradle.parent` (of type `Gradle`).

The solution we found is to use `gradle.startParameter.currentDir` and then search the hierarchy upwards for a `settings.gradle(.kts)` file. This is re-implementing what Gradle does to determine `BuildLayout.settingsDirectory`.

Maybe another feature to consider in this area would be something like `Gradle.layout` that gives you early access to `settingsDirectory`?

**Comment by h0tk3y** [comment]
In addition to `new File("relative-path")`, there is some more API in the JDK that would implicitly use the CWD to construct a relative path.

At least `FileInputStream`, `FileOutputStream`, `FileReader` etc. have string-path constructors that would do this.

**Comment by Vampire** [comment]
Also the NIO stuff like `Path.of` and `Paths.get`, ...

**Comment by eskatos** [comment]
Just a note that `--foreground` currently runs the daemon in the client JVM. This would need to be changed to run a dedicated daemon process.
